File: src/main/java/com/ada/shared/infrastructure/DataDirectoryInitializer.java
package com.ada.shared.infrastructure;

import java.nio.file.*;
import java.util.*;
import lombok.RequiredArgsConstructor;
import org.springframework.boot.ApplicationRunner;
import org.springframework.context.annotation.*;

@Configuration
@RequiredArgsConstructor
public class DataDirectoryInitializer {
  private final AdaProperties properties;

  @Bean
  public ApplicationRunner initializeDataDirectory() {
    return args -> {
      for (String d : List.of("db", "logs", "backups", "exports", "models", "runtime"))
        Files.createDirectories(properties.getNormalizedDataDirectory().resolve(d));
    };
  }
}

File: src/main/kotlin/com/ada/shared/infrastructure/DataDirectoryInitializer.kt
package com.ada.shared.infrastructure

import org.springframework.boot.ApplicationRunner
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import java.nio.file.Files

@Configuration
class DataDirectoryInitializer(
    private val properties: AdaProperties,
) {
    /**
     * Creates an application runner that initializes the required data subdirectories at startup.
     *
     * @return An application runner that creates the `db`, `logs`, `backups`, `exports`, `models`, and `runtime` directories.
     */
    @Bean
    fun initializeDataDirectory(): ApplicationRunner = ApplicationRunner {
        val dataDirectory = properties.normalizedDataDirectory
        listOf("db", "logs", "backups", "exports", "models", "runtime")
            .map(dataDirectory::resolve)
            .forEach(Files::createDirectories)
    }
}
